src/stream.rs

```diff
@@ -1,599 +1,422 @@
+// Copyright 2021 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+//! Streaming decryption functionality for memory-efficient processing of large encrypted files.
+
 use crate::{
-    decrypt::decrypt_chunk,
-    encrypt::encrypt_chunk,
-    error::{Error, Result},
-    utils::{extract_hashes, get_pki},
-    DataMap, EncryptedChunk,
+    decrypt::decrypt_chunk, get_root_data_map_parallel, utils::extract_hashes, ChunkInfo, DataMap,
+    Result,
 };
 use bytes::Bytes;
-use std::{
-    collections::BTreeMap,
-    fs::{File, OpenOptions},
-    io::{Error as IoError, ErrorKind},
-    io::{Read, Seek, SeekFrom, Write},
-    path::PathBuf,
-};
-use tempfile::{tempdir, TempDir};
 use xor_name::XorName;
 
-/// The streaming encryptor to carry out the encryption on fly, chunk by chunk.
-#[derive(Clone)]
-pub struct StreamSelfEncryptor {
-    // File path for the encryption target.
-    file_path: PathBuf,
-    // List of `(start_position, end_position)` for each chunk for the target file.
-    batch_positions: Vec<(usize, usize)>,
-    // Current step (i.e. chunk_index) for encryption
-    chunk_index: usize,
-    // Progressing DataMap
-    data_map: Vec<crate::ChunkInfo>,
-    // Progressing collection of source chunks' names
-    src_hashes: BTreeMap<usize, XorName>,
-    // File path to flush encrypted_chunks into.
-    chunk_dir: Option<PathBuf>,
+/// Iterator that yields decrypted chunks as `Bytes` in streaming fashion.
+///
+/// This provides memory-efficient decryption by processing chunks in batches
+/// and yielding them one at a time without buffering the entire file.
+pub struct StreamingDecrypt<F> {
+    chunk_infos: Vec<ChunkInfo>,
+    src_hashes: Vec<XorName>,
+    get_chunk_parallel: F,
+    current_batch_start: usize,
+    current_batch_chunks: Vec<Bytes>,
+    current_batch_index: usize,
+    batch_size: usize,
 }
 
-impl StreamSelfEncryptor {
```

Comment:
The PR just provides a non-disk-file dependent streaming decryptor.

Prefer leave StreamSelfEncryptor unless there is a new replacement presents,
to avoid the feature got `forgotten` .